File: Samples/DotNet/C#/Diagnostic/VccCcInterrupt/VccCcInterrupt.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace VccCcInterrupt
{
    public partial class VccCcInterrupt : Form
    {
        /// <summary>
        /// Handle to the first board
        /// </summary>
        PCI1500_NET.PCI1500 board;

        /// Delegate that enables to add an event in the box in another thread
        private delegate void AddEventDelegate(String eventInfo);

        /// <summary>
        /// Constructor of the form
        /// </summary>
        public VccCcInterrupt()
        {
            InitializeComponent();
        }

        /// <summary>
        /// Add an event in the event box
        /// </summary>
        /// <param name="eventInfo"></param>
        private void AddEvent(String eventInfo)
        {
            eventBox.Text += DateTime.Now.ToString() + " : " + eventInfo + "\r\n";
            eventBox.SelectionStart = eventBox.Text.Length;
            eventBox.ScrollToCaret();
        }

        /// <summary>
        /// Call at the first load
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void VccCcInterrupt_Load(object sender, EventArgs e)
        {
            int ret = 0;
            // Open the first board of the system
            this.board = PCI1500_NET.PCI1500.OpenBoard(0);
            // Error handling
            if (this.board == null)
            {
                MessageBox.Show("Error OpenBoard returned null");
                this.Close();
                return;
            }

            // Listen to the diagnostic event
            this.board.DiagnosticEventReceived += new PCI1500_NET.PCI1500.DiagnosticEventHandler(board_DiagnosticEventReceived);

            // Set all the output to high (to generate output interrupt)
            if ((ret = this.board.Set16DigitalOutputsOn(0xFFFF)) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in Set16DigitalOutputsOn");
                this.Close();
                return;
            }
            // Set the interrupt routine
            if ((ret = this.board.SetBoardIntRoutine()) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in SetBoardIntRoutine");
                this.Close();
                return;
            }
            // Enable the event logic on port 1 in order to receive CC and VCC diagnostic
            if ((ret = this.board.InitEventLogic(1, 0, "xxxxxx")) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in InitEventLogic");
                this.Close();
                return;
            }
            // Start the event logic
            if ((ret = this.board.StartEventLogic(1)) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in StartEventLogic");
                this.Close();
                return;
            }
        }

        /// <summary>
        /// Occurs when we get a diagnostic event
        /// </summary>
        /// <param name="source"></param>
        /// <param name="args"></param>
        void board_DiagnosticEventReceived(object source, PCI1500_NET.DiagnosticEventArgs args)
        {
            if (args.source == 0)
            {
                this.Invoke(new AddEventDelegate(this.AddEvent), "Short-circuit occured");
            }
            else if (args.source == 1)
            {
                this.Invoke(new AddEventDelegate(this.AddEvent), "Supply voltage lost");
            }
        }

        /// <summary>
        /// Occurs when the form is being closed
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void VccCcInterrupt_FormClosing(object sender, FormClosingEventArgs e)
        {
            int ret = 0;
            // First set the digital output memory to on
            if ((ret = this.board.SetDigitalOutputMemoryOn()) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in SetDigitalOutputMemoryOn");
            }
            // Then set all the output to off
            if ((ret = this.board.Set16DigitalOutputsOff(0xFFFF)) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in Set16DigitalOutputsOff");
            }
            // Reset the digital output memory to off 
            if ((ret = this.board.SetDigitalOutputMemoryOff()) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in SetDigitalOutputMemoryOff");
            }
            // Stop the event logic
            if ((ret = this.board.StopEventLogic(1)) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in StopEventLogic");
            }
            // Reset the board interrupt routine
            if ((ret = this.board.ResetBoardIntRoutine()) != 0)
            {
                MessageBox.Show("Error " + ret.ToString() + " in ResetBoardIntRoutine");
            }
        }
    }
}

File: Samples/C/Diagnostic/VccCcInterrupt/VccCcInterrupt.cpp
/**
* Demonstrates how to use the short-circuit / voltage supply lost detection
**/
#include "stdafx.h"

int _tmain ( int /*argc*/, _TCHAR* /*argv[]*/ )
{
	INT returnValue = 0;
	HANDLE boardHandle = NULL;

	/* call the util method to open a board */
	returnValue = OpenBoard(&boardHandle);
	if (returnValue != 0)
	{
		printf("Error while opening the board : %d \n", (int) returnValue);
		_getch();
		return -1;
	}

	/* display some useful informations */
	printf("* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * \n");
	printf("* Short circuit - Voltage supply lost sample. \n");
	printf("* Enables to detect short circuit or voltage supply lost thanks to\n");
	printf("* interrupts.\n");
	printf("* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * \n\n");

	/* set the digital output memory */
	returnValue = i_PCI1500_SetDigitalOutputMemoryOn(boardHandle);
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_SetDigitalOutputMemoryOn \n", returnValue);
		i_PCI1500_CloseBoard(boardHandle);
		_getch();
		return -1;
	}
	printf("i_PCI1500_SetDigitalOutputMemoryOn : successful\n");

	/* set the interrupt routine */
	returnValue = i_PCI1500_SetBoardIntRoutine(boardHandle, InterruptRoutine);
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_SetBoardIntRoutine \n", returnValue);
		i_PCI1500_CloseBoard(boardHandle);
		_getch();
		return -1;
	}
	printf("i_PCI1500_SetBoardIntRoutine : successful\n");

	/* enables the event logic on port 1 to receive the vcc and cc interruption */
	returnValue = i_PCI1500_InitEventLogic(boardHandle,PCI1500_PORT_1,PCI1500_LOGIC_OR,"xxxxxx");
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_InitEventLogic \n", returnValue);
		i_PCI1500_ResetBoardIntRoutine(boardHandle);
		i_PCI1500_CloseBoard(boardHandle);
		_getch();
		return -1;
	}
	printf("i_PCI1500_InitEventLogic : successful\n");

	/* start the event logic */
	returnValue = i_PCI1500_StartEventLogic(boardHandle,PCI1500_PORT_1);
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_StartEventLogic \n", returnValue);
		i_PCI1500_ResetBoardIntRoutine(boardHandle);
		i_PCI1500_CloseBoard(boardHandle);
		_getch();
		return -1;
	}
	printf("i_PCI1500_StartEventLogic : successful\n");

	/* We set the digital outputs to ON in order to be able to short-circuit them */
	returnValue = i_PCI1500_Set16DigitalOutputsOn( boardHandle, 0xFFFF );
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_Set16DigitalOutputsOn \n", returnValue);
		i_PCI1500_StopEventLogic(boardHandle, PCI1500_PORT_1);
		i_PCI1500_ResetBoardIntRoutine(boardHandle);
		i_PCI1500_CloseBoard(boardHandle);
		_getch();
		return -1;
	}
	printf("i_PCI1500_Set16DigitalOutputsOn : successful\n");

	printf("The digital outputs are ON. \n");
	printf("You can short-circuit them to get a CC interrupt.\n");
	printf("You can lower the input voltage to get a VCC interrupt. \n");
	printf("Press space to set the outputs again. \n");
	printf("Press any other key to stop the sample.\n");

	/* endless loop */
	for(;;)
	{
		Sleep(100);
		if ( _kbhit() )
		{ 
			/* space key ? */
			if ( _getch() == 32 )
			{
				returnValue = i_PCI1500_Set16DigitalOutputsOn( boardHandle, 0xFFFF );
				if ( returnValue != 0 )
				{
					printf("Error %d in i_PCI1500_Set16DigitalOutputsOn \n", returnValue);
					i_PCI1500_StopEventLogic(boardHandle, PCI1500_PORT_1);
					i_PCI1500_ResetBoardIntRoutine(boardHandle);
					i_PCI1500_CloseBoard(boardHandle);
					_getch();
					return -1;
				}
				printf("i_PCI1500_Set16DigitalOutputsOn : successful\n");
			}
			/* other key */
			else
			{
				break;
			}
		}
	}

	/* reset the digital outputs */
	returnValue = i_PCI1500_Set16DigitalOutputsOff( boardHandle, 0xFFFF );
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_Set16DigitalOutputsOff \n", returnValue);
	}
	else
	{
		printf("i_PCI1500_Set16DigitalOutputsOff : successful\n");
	}

	/* stop the event logic */
	returnValue = i_PCI1500_StopEventLogic(boardHandle, PCI1500_PORT_1);
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_StopEventLogic \n", returnValue);
	}
	else
	{
		printf("i_PCI1500_StopEventLogic : successful\n");
	}

	/* reset the interrupt routine */
	returnValue = i_PCI1500_ResetBoardIntRoutine(boardHandle);
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_ResetBoardIntRoutine \n", returnValue);
	}
	else
	{
		printf("i_PCI1500_ResetBoardIntRoutine : successful\n");
	}

	/* release the board */
	returnValue = i_PCI1500_CloseBoard(boardHandle);
	if ( returnValue != 0 )
	{
		printf("Error %d in i_PCI1500_CloseBoard \n", returnValue);
	}
	else
	{
		printf("i_PCI1500_CloseBoard : successful\n");
	}

	/* exit */
	printf("\n Press any key to exit the sample ...");
	_getch();
	return 0;
}

